gg_fit_curve.py

```diff
@@ -0,0 +1,402 @@
+    C = [[0.0, 0.0], [0.0, 0.0]]
+    X = [0.0, 0.0]
+
+    for i in range(num_pts):
+        C[0][0] += Vector.dot_product(A[i][0], A[i][0])
+        C[0][1] += Vector.dot_product(A[i][0], A[i][1])
+
+        C[1][0] = C[0][1]
+        C[1][1] += Vector.dot_product(A[i][1], A[i][1])
+
+        vec_tmp = points[first + i].sub(
+            points[first]
+            .scale(_B0(u_prime[i]))
+            .add(
+                points[first]
+                .scale(_B1(u_prime[i]))
+                .add(
+                    points[last]
+                    .scale(_B2(u_prime[i]))
+                    .add(points[last].scale(_B3(u_prime[i])))
+                )
+            )
+        )
+
+        X[0] += Vector.dot_product(A[i][0], vec_tmp)
+        X[1] += Vector.dot_product(A[i][1], vec_tmp)
+
+    # Compute the determinants of C and X
+    det_C0_C1 = C[0][0] * C[1][1] - C[1][0] * C[0][1]
+    det_C0_X = C[0][0] * X[1] - C[1][0] * X[0]
+    det_X_C1 = X[0] * C[1][1] - X[1] * C[0][1]
+
+    # Finally, derive alpha values
+    alpha_l = alpha_r = 0.0
+    if det_C0_C1 != 0.0:
+        alpha_l = det_X_C1 / det_C0_C1
+        alpha_r = det_C0_X / det_C0_C1
+
+    # If alpha negative, use the Wu/Barsky heuristic (see text)
+    # (if alpha is 0, you get coincident control points that lead to
+    # divide by zero in any subsequent NewtonRaphsonRootFind() call.
+    seg_length = Vector.p0_to_p1(points[first], points[last]).length()
+    epsilon = 1.0e-6 * seg_length
+    if alpha_l < epsilon or alpha_r < epsilon:
+        # fall back on standard (probably inaccurate) formula, and subdivide further if needed.
+        dist = seg_length / 3.0
+        return (
+            points[first],
+            points[first].add(tan_left.scale(dist)),
+            points[last].add(tan_right.scale(dist)),
+            points[last],
+        )
+
+    #  First and last control points of the Bezier curve are
+    #  positioned exactly at the first and last data points
+    #  Control points 1 and 2 are positioned an alpha distance out
+    #  on the tangent vectors, left and right, respectively
+    return (
+        points[first],
+        points[first].add(tan_left.scale(alpha_l)),
+        points[last].add(tan_right.scale(alpha_r)),
+        points[last],
+    )
+
+
+# ChordLengthParameterize
+#    Assign parameter values to digitized points
+#    using relative distances between points.
+def _chord_length_parameterize(points, first, last) -> Sequence[float]:
+    u = [0.0] * (last - first + 1)
+    for i in range(first + 1, last + 1):
+        u[i - first] = (
+            u[i - first - 1] + Vector.p0_to_p1(points[i - 1], points[i]).length()
+        )
+
+    for i in range(first + 1, last + 1):
+        u[i - first] = u[i - first] / u[last - first]
+
+    return u
+
+
+# TODO better "u" handling
+def _max_error(points, first, last, curve, u) -> Tuple[float, int]:
+    assert degree(curve) == 3
+    max_err = 0.0
+    split_at = int((last - first + 1) / 2)
```

Comment:
for integer division you could use the `//` operator, which similarly truncates
```suggestion
    split_at = (last - first + 1) // 2
```